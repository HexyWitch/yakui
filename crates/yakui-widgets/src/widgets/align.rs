use yakui_core::dom::Dom;
use yakui_core::geometry::{Constraints, Vec2};
use yakui_core::layout::LayoutDom;
use yakui_core::paint::PaintDom;
use yakui_core::widget::Widget;
use yakui_core::{Alignment, Response};

use crate::util::widget_children;

#[derive(Debug, Clone)]
pub struct Align {
    pub alignment: Alignment,
}

impl Align {
    pub fn new(alignment: Alignment) -> Self {
        Self { alignment }
    }

    pub fn center() -> Self {
        Self {
            alignment: Alignment::CENTER,
        }
    }

    pub fn show<F: FnOnce()>(self, children: F) -> Response<AlignWidget> {
        widget_children::<AlignWidget, F>(children, self)
    }
}

#[derive(Debug)]
pub struct AlignWidget {
    props: Align,
}

pub type AlignResponse = ();

impl Widget for AlignWidget {
    type Props = Align;
    type Response = AlignResponse;

    fn new(props: Self::Props) -> Self {
        Self { props }
    }

    fn update(&mut self, props: Self::Props) {
        self.props = props;
    }

    fn layout(&self, dom: &Dom, layout: &mut LayoutDom, input: Constraints) -> Vec2 {
        let node = dom.get_current();

        // FIXME: Need to behave when given unbound constraints
        let self_size = input.max;
        let align = self.props.alignment.as_vec2();

        for &child in &node.children {
            let child_size = layout.calculate(dom, child, input);
            layout.set_pos(child, align * self_size - align * child_size);
        }

        self_size
    }

    fn paint(&self, dom: &Dom, layout: &LayoutDom, paint: &mut PaintDom) {
        let node = dom.get_current();
        for &child in &node.children {
            paint.paint(dom, layout, child);
        }
    }

    fn respond(&mut self) -> Self::Response {}
}
