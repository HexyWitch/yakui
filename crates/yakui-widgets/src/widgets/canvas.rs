use yakui_core::widget::{PaintContext, Widget};
use yakui_core::Response;

use crate::ignore_debug::IgnoreDebug;
use crate::util::widget;

type DrawCallback = Box<dyn Fn(PaintContext<'_>) + 'static>;

/**
Allows the user to draw arbitrary graphics in a region.

Responds with [CanvasResponse].
*/
#[derive(Debug)]
pub struct Canvas {
    draw: IgnoreDebug<Option<DrawCallback>>,
}

impl Canvas {
    pub fn new(draw: impl Fn(PaintContext<'_>) + 'static) -> Self {
        Self {
            draw: IgnoreDebug(Some(Box::new(draw))),
        }
    }

    pub fn show(self) -> Response<CanvasWidget> {
        widget::<CanvasWidget>(self)
    }
}

#[derive(Debug)]
pub struct CanvasWidget {
    props: Canvas,
}

pub type CanvasResponse = ();

impl Widget for CanvasWidget {
    type Props = Canvas;
    type Response = CanvasResponse;

    fn new() -> Self {
        Self {
            props: Canvas {
                draw: IgnoreDebug(None),
            },
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;
    }

    fn paint(&self, ctx: PaintContext<'_>) {
        if let IgnoreDebug(Some(draw)) = &self.props.draw {
            (draw)(ctx);
        }
    }
}
