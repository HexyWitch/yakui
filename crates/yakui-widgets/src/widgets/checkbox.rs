use yakui_core::dom::Dom;
use yakui_core::event::{EventInterest, EventResponse, WidgetEvent};
use yakui_core::geometry::{Constraints, Vec2};
use yakui_core::input::MouseButton;
use yakui_core::layout::LayoutDom;
use yakui_core::paint::{PaintDom, PaintRect};
use yakui_core::widget::Widget;
use yakui_core::Response;

use crate::colors;

const OUTER_SIZE: f32 = 32.0;
const INNER_SIZE: f32 = 24.0;

#[derive(Debug)]
#[non_exhaustive]
pub struct Checkbox {
    pub checked: bool,
}

impl Checkbox {
    pub fn new(checked: bool) -> Self {
        Self { checked }
    }

    pub fn show(self) -> Response<CheckboxWidget> {
        crate::util::widget::<CheckboxWidget>(self)
    }
}

#[derive(Debug)]
pub struct CheckboxWidget {
    props: Checkbox,
    hovering: bool,
    mouse_down: bool,
    just_toggled: bool,
}

#[derive(Debug)]
#[non_exhaustive]
pub struct CheckboxResponse {
    pub checked: bool,
}

impl Widget for CheckboxWidget {
    type Props = Checkbox;
    type Response = CheckboxResponse;

    fn new(props: Self::Props) -> Self {
        Self {
            props,
            hovering: false,
            mouse_down: false,
            just_toggled: false,
        }
    }

    fn update(&mut self, props: Self::Props) {
        self.props = props;
    }

    fn respond(&mut self) -> Self::Response {
        let mut checked = self.props.checked;
        if self.just_toggled {
            checked = !checked;
            self.just_toggled = false;
        }

        CheckboxResponse { checked }
    }

    fn paint(&self, dom: &Dom, layout: &LayoutDom, paint: &mut PaintDom) {
        let layout_node = layout.get(dom.current()).unwrap();
        let outer_rect = layout_node.rect.div_vec2(layout.viewport().size());

        let padding = Vec2::splat(OUTER_SIZE - INNER_SIZE);
        let mut check_rect = layout_node.rect;
        check_rect.set_pos(check_rect.pos() + padding / 2.0);
        check_rect.set_size(check_rect.size() - padding);
        check_rect = check_rect.div_vec2(layout.viewport().size());

        let mut bg = PaintRect::new(outer_rect);
        bg.color = colors::BACKGROUND_3;
        paint.add_rect(bg);

        if self.props.checked {
            crate::icons::cross(paint, check_rect, colors::TEXT);
        }
    }

    fn layout(&self, _dom: &Dom, _layout: &mut LayoutDom, constraints: Constraints) -> Vec2 {
        constraints.constrain(Vec2::splat(OUTER_SIZE))
    }

    fn event_interest(&self) -> EventInterest {
        EventInterest::MOUSE
    }

    fn event(&mut self, event: &WidgetEvent) -> EventResponse {
        match event {
            WidgetEvent::MouseEnter => {
                self.hovering = true;
                EventResponse::Sink
            }
            WidgetEvent::MouseLeave => {
                self.hovering = false;
                EventResponse::Sink
            }
            WidgetEvent::MouseButtonChanged(MouseButton::One, down) => {
                if *down {
                    self.mouse_down = true;
                    EventResponse::Sink
                } else if self.mouse_down {
                    self.mouse_down = false;
                    self.just_toggled = true;
                    EventResponse::Sink
                } else {
                    EventResponse::Bubble
                }
            }
            WidgetEvent::MouseButtonChangedOutside(MouseButton::One, _) => {
                self.mouse_down = false;
                EventResponse::Bubble
            }
            _ => EventResponse::Bubble,
        }
    }
}
