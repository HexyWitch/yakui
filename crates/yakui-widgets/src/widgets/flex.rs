use yakui_core::geometry::FlexFit;
use yakui_core::widget::Widget;
use yakui_core::Response;

use crate::util::widget_children;

#[derive(Debug)]
pub struct Flexible {
    pub flex: u32,
    pub fit: FlexFit,
}

impl Flexible {
    pub fn new(flex: u32) -> Self {
        Self {
            flex,
            fit: FlexFit::Loose,
        }
    }

    pub fn expanded() -> Self {
        Self {
            flex: 1,
            fit: FlexFit::Tight,
        }
    }

    pub fn show<F: FnOnce()>(self, children: F) -> Response<FlexWidget> {
        widget_children::<FlexWidget, F>(children, self)
    }
}

#[derive(Debug)]
pub struct FlexWidget {
    props: Flexible,
}

pub type FlexResponse = ();

impl Widget for FlexWidget {
    type Props = Flexible;
    type Response = FlexResponse;

    fn new() -> Self {
        Self {
            props: Flexible::new(0),
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;
    }

    fn flex(&self) -> (u32, FlexFit) {
        (self.props.flex, self.props.fit)
    }
}
