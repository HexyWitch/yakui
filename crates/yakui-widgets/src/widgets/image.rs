use yakui_core::dom::Dom;
use yakui_core::geometry::{Color3, Constraints, Rect, Vec2};
use yakui_core::layout::LayoutDom;
use yakui_core::paint::{PaintDom, PaintRect};
use yakui_core::widget::Widget;
use yakui_core::{Response, TextureId};

use crate::util::widget;

/**
Displays an image.

Responds with [ImageResponse].
*/
#[derive(Debug, Clone)]
#[non_exhaustive]
pub struct Image {
    pub image: Option<TextureId>,
    pub size: Vec2,
}

impl Image {
    pub fn new(image: TextureId, size: Vec2) -> Self {
        Self {
            image: Some(image),
            size,
        }
    }

    pub fn show(self) -> Response<ImageWidget> {
        widget::<ImageWidget>(self)
    }
}

#[derive(Debug)]
pub struct ImageWidget {
    props: Image,
}

pub type ImageResponse = ();

impl Widget for ImageWidget {
    type Props = Image;
    type Response = ImageResponse;

    fn new() -> Self {
        Self {
            props: Image {
                image: None,
                size: Vec2::ZERO,
            },
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;
    }

    fn layout(&self, _dom: &Dom, _layout: &mut LayoutDom, input: Constraints) -> Vec2 {
        input.constrain(self.props.size)
    }

    fn paint(&self, dom: &Dom, layout: &LayoutDom, output: &mut PaintDom) {
        let layout_node = layout.get(dom.current()).unwrap();

        if let Some(image) = self.props.image {
            let mut rect = PaintRect::new(layout_node.rect);
            rect.color = Color3::WHITE;
            rect.texture = Some((image, Rect::ONE));
            output.add_rect(rect);
        }
    }
}
