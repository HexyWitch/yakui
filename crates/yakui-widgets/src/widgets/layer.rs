use yakui_core::widget::{PaintContext, Widget};
use yakui_core::Response;

use crate::util::widget_children;

/**
Creates a new layer that will draw over the top of the current layer.
*/
#[derive(Debug, Clone)]
#[non_exhaustive]
pub struct Layer {}

impl Layer {
    pub fn new() -> Self {
        Self {}
    }

    pub fn show<F: FnOnce()>(self, children: F) -> Response<LayerWidget> {
        widget_children::<LayerWidget, F>(children, self)
    }
}

#[derive(Debug)]
pub struct LayerWidget {
    props: Layer,
}

pub type LayerResponse = ();

impl Widget for LayerWidget {
    type Props = Layer;
    type Response = LayerResponse;

    fn new() -> Self {
        Self { props: Layer {} }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;
    }

    fn paint(&self, mut ctx: PaintContext<'_>) {
        ctx.paint.push_layer();

        let node = ctx.dom.get_current();
        for &child in &node.children {
            ctx.paint(child);
        }

        ctx.paint.pop_layer();
    }
}
