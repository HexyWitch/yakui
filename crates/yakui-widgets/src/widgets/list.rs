use yakui_core::dom::Dom;
use yakui_core::geometry::{Constraints, FlexFit, Vec2};
use yakui_core::layout::LayoutDom;
use yakui_core::widget::Widget;
use yakui_core::{CrossAxisAlignment, Direction, MainAxisAlignment, MainAxisSize, Response};

use crate::util::widget_children;

/**
Lays out children in a single direction. Supports flex sizing.

This is one of the most common and useful layout widgets.

Responds with [ListResponse].

Shorthand:
```rust
# use yakui_widgets::doc_yakui as yakui;
# let _handle = yakui_widgets::DocTest::start();
yakui::column(|| {
    yakui::label("on top");
    yakui::label("on bottom");
});

yakui::row(|| {
    yakui::label("left");
    yakui::label("right");
});
```
*/
#[derive(Debug, Clone)]
#[non_exhaustive]
pub struct List {
    pub direction: Direction,
    pub main_axis_size: MainAxisSize,
    pub main_axis_alignment: MainAxisAlignment,
    pub cross_axis_alignment: CrossAxisAlignment,
}

impl List {
    pub fn new(direction: Direction) -> Self {
        Self {
            direction,
            main_axis_size: MainAxisSize::Max,
            main_axis_alignment: MainAxisAlignment::Start,
            cross_axis_alignment: CrossAxisAlignment::Start,
        }
    }

    pub fn vertical() -> Self {
        Self::new(Direction::Down)
    }

    pub fn horizontal() -> Self {
        Self::new(Direction::Right)
    }

    pub fn show<F: FnOnce()>(self, children: F) -> Response<ListWidget> {
        widget_children::<ListWidget, F>(children, self)
    }
}

#[derive(Debug)]
pub struct ListWidget {
    props: List,
}

pub type ListResponse = ();

impl Widget for ListWidget {
    type Props = List;
    type Response = ListResponse;

    fn new() -> Self {
        Self {
            props: List::horizontal(),
        }
    }

    fn update(&mut self, props: Self::Props) -> Self::Response {
        self.props = props;
    }

    // This approach to layout is based on Flutter's Flex layout algorithm.
    //
    // https://api.flutter.dev/flutter/widgets/Flex-class.html#layout-algorithm
    fn layout(&self, dom: &Dom, layout: &mut LayoutDom, input: Constraints) -> Vec2 {
        let node = dom.get_current();
        let direction = self.props.direction;

        let mut total_main_axis_size = 0.0;
        let mut max_cross_axis_size = 0.0;

        // First, we lay out any children that do not flex, giving them unbound
        // main axis constraints. This ensures that we don't unfairly squish
        // later widgets in the layout.
        //
        // Simultaneously, we'll track the total value of all flexible elements
        // so that we can divide the remaining space up later.
        let mut total_flex = 0;
        for &child_index in &node.children {
            let child = dom.get(child_index).unwrap();
            let (flex, _fit) = child.widget.flex();
            total_flex += flex;

            if flex != 0 {
                continue;
            }

            let constraints = Constraints {
                min: Vec2::ZERO,
                max: direction.vec2(f32::INFINITY, direction.get_cross_axis(input.max)),
            };

            let size = layout.calculate(dom, child_index, constraints);
            total_main_axis_size += direction.get_main_axis(size);
            max_cross_axis_size = f32::max(max_cross_axis_size, direction.get_cross_axis(size));
        }

        // Next, lay out all flexible elements, giving them each some portion of
        // the remaining space based on their flex factor.
        let remaining_main_axis =
            (direction.get_main_axis(input.max) - total_main_axis_size).max(0.0);
        for &child_index in &node.children {
            let child = dom.get(child_index).unwrap();
            let (flex, fit) = child.widget.flex();

            if flex == 0 {
                continue;
            }

            let main_axis_size = flex as f32 * remaining_main_axis / total_flex as f32;

            // The maximum main axis size is based on the portion of the
            // remaining space that we allocated to this widget.
            //
            // We pass along the maximum cross axis size from our incoming
            // constraints.
            let max = direction.vec2(main_axis_size, direction.get_cross_axis(input.max));
            let constraints = match fit {
                FlexFit::Loose => Constraints::loose(max),
                FlexFit::Tight => Constraints {
                    min: direction.vec2(main_axis_size, 0.0),
                    max,
                },
            };

            let size = layout.calculate(dom, child_index, constraints);
            total_main_axis_size += direction.get_main_axis(size);
            max_cross_axis_size = f32::max(max_cross_axis_size, direction.get_cross_axis(size));
        }

        // Finally, position all children based on the sizes calculated above.
        let mut next_pos = Vec2::ZERO;
        for &child_index in &node.children {
            let child_layout = layout.get_mut(child_index).unwrap();
            child_layout.rect.set_pos(next_pos);

            next_pos += direction.only_main_axis(child_layout.rect.size());
        }

        input.constrain(direction.vec2(total_main_axis_size, max_cross_axis_size))
    }
}
