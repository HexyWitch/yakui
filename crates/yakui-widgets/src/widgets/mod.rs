mod align;
mod button;
mod checkbox;
mod colored_box;
mod constrained_box;
mod flexible;
mod image;
mod list;
mod pad;
mod text;
mod textbox;
mod window;

pub use self::align::*;
pub use self::button::*;
pub use self::checkbox::*;
pub use self::colored_box::*;
pub use self::constrained_box::*;
pub use self::flexible::*;
pub use self::image::*;
pub use self::list::*;
pub use self::pad::*;
pub use self::text::*;
pub use self::textbox::*;
pub use self::window::*;
