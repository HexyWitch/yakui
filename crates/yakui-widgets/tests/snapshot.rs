use yakui::MainAxisSize;
use yakui_core::geometry::Color3;
use yakui_core::Alignment;
use yakui_test::{run, Test};
use yakui_widgets::widgets::List;

#[test]
fn colored_box() {
    run!({
        yakui_widgets::colored_box(Color3::RED, [50.0, 50.0]);
    });
}

#[test]
fn column() {
    run!({
        yakui_widgets::column(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn row() {
    run!({
        yakui_widgets::row(|| {
            rect_50x50();
            rect_50x50();
            rect_50x50();
        });
    });
}

#[test]
fn column_min() {
    run!({
        let mut column = List::column();
        column.main_axis_size = MainAxisSize::Min;
        column.show(|| {
            rect_50x50();
            rect(100, 50);
            rect(50, 100);
        });
    });
}

#[test]
fn align_center() {
    run!({
        yakui_widgets::center(|| {
            rect_50x50();
        });
    });
}

#[test]
fn align_bottom_right() {
    run!({
        yakui_widgets::align(Alignment::BOTTOM_RIGHT, || {
            rect_50x50();
        });
    });
}

fn rect<V: Dim>(w: V, h: V) {
    yakui_widgets::colored_box(Color3::WHITE, [w.to_f32(), h.to_f32()]);
}

fn rect_50x50() {
    rect(50.0, 50.0);
}

trait Dim {
    fn to_f32(self) -> f32;
}

impl Dim for f32 {
    fn to_f32(self) -> f32 {
        self
    }
}

impl Dim for i32 {
    fn to_f32(self) -> f32 {
        self as f32
    }
}
