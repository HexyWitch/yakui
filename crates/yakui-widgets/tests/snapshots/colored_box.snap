---
source: crates/yakui-widgets/tests/snapshot.rs
assertion_line: 9
expression: view
---
- ColoredBoxWidget pos(0, 0) size(1000, 1000)

